typed-racket-lib/typed-racket/typecheck/tc-funapp.rkt

```diff
@@ -160,9 +161,9 @@
       [(? resolvable?)
        (tc/funapp f-stx args-stx (resolve-once f-type) args-res expected)]
       ;; a union of functions can be applied if we can apply all of the elements
-      [(Union: (and ts (list (? Function?) ...)))
+      [(Union: ts) #:when (for/and ([t (in-hset ts)]) (Function? t))
```

Comment:
We could probably avoid traversing the hset twice. Maybe not worth it right now, but something to think about.